elastic/logstash
# Added  metric to count removed events when age retention kicks in

## Release notes

Adds the metric `expired_events` counter to track the events removed by age retention policy.

## What does this PR do?

Adds the metric measure `expired_events` to node stats graph to track the number of events removed by age retention pruning policy.
In the `_node/stats` API, under the path `pipelines.<pipeline upstream>.dead_letter_queue` enrich the document with the ne metric, so it results as:
```json
{
    "queue_size_in_bytes": 3127024,
    "storage_policy": "drop_newer",
    "dropped_events": 0,
    "max_queue_size_in_bytes": 1073741824,
    "expired_events": 50056,
    "last_error": "no errors"
}
```
The number of expired events is incremented every time a segment file is deleted; this happens opening the file and counting the `complete` and `start` records.

## Why is it important/What is the impact to the user?
Permit to the user to monitor the expired events from DLQ. Expired events could be or not be processed. Is a measure of events removed where is not certain if they were processed from the downstream DLQ pipeline.

## Checklist



- ~~[ ] I have made corresponding changes to the documentation~~ follow up PR

## Author's Checklist



## How to test this PR locally
- run a test case as described in #14255
- do a call to HTTP monitoring API
```
GET localhost:9600/_node/stats
```
and check the `expired_event` counter

## Related issues


- Relates #10493

## Use cases

As user I want to monitor how many events are dropped from the DLQ because of age retention policy cleaning.
